src/main/java/spark/servlet/SparkServlet.java

```diff
@@ -0,0 +1,273 @@
+            path += SLASH;
+        }
+
+        if (path.startsWith(filterPath)) {
+            path = path.substring(filterPath.length());
+        }
+
+        if (!path.startsWith(SLASH)) {
+            path = SLASH + path;
+        }
+
+        try {
+            path = URLDecoder.decode(path, "UTF-8");
+        } catch (UnsupportedEncodingException ignore) {
+            // this can't really ever happen
+        }
+
+        LOGGER.debug("Relative path = {}", path);
+        return path;
+    }
+
+    private void populateWrapperSupplier(ServletConfig config) {
+        // Do not override an injected supplier through initialization
+        if (requestSupplier != null) {
+            return;
+        }
+
+        String wrapperSupplierName = config.getInitParameter(WRAPPER_SUPPLIER_PARAM_NAME);
+
+        if (StringUtils.isNotBlank(wrapperSupplierName)) {
+            try {
+                Class<?> wrapperClass = Class.forName(wrapperSupplierName);
+                if (BiFunction.class.isAssignableFrom(wrapperClass)) {
+                    requestSupplier =
+                            (BiFunction<HttpServletRequest, String, HttpServletRequestWrapper>) wrapperClass.newInstance();
+                }
+            } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+                LOGGER.debug(
+                        "Error converting {} to BiFunction<HttpServletRequest, String, HttpServletRequestWrapper>; "
+                                + "falling back to default",
+                        wrapperSupplierName,
+                        e);
+            }
+
+        }
+
+        if (requestSupplier == null) {
+            requestSupplier = DEFAULT_REQ_FUNC;
+        }
+    }
+
+    private void populateSparkApplications(ServletConfig config) {
+        // Do not override injected spark applications through initialization
+        if (!CollectionUtils.isEmpty(sparkApplications)) {
+            return;
+        }
+
+        String applications = config.getInitParameter(APPLICATION_CLASS_PARAM);
+
+        if (StringUtils.isNotBlank(applications)) {
+            sparkApplications.addAll(Pattern.compile(",")
+                    .splitAsStream(applications)
+                    .map(String::trim)
+                    .map(this::getApplication)
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toList()));
+        }
+    }
+
+    private SparkApplication getApplication(String applicationClassName) {
+        try {
+            Class<?> appClass = Class.forName(applicationClassName);
+            if (SparkApplication.class.isAssignableFrom(appClass)) {
+                return SparkApplication.class.cast(appClass.newInstance());
+            }
+        } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
+            LOGGER.debug("Error converting {} to SparkApplication", applicationClassName, e);
+        }
+
```

Comment:
This is way lenient - personally I'd prefer both the `isAssignableFrom()` check and caught exceptions to raise hell to just going on and merely logging at debug level. Being more strict would also get rid of the potential null return value and `Objects.nonNull()` that handles it.